File: client/src/game/protocol.ts
export type ClientMessage =
  | { type: 'join' }
  | { type: 'ping'; clientTime: number }
  | { type: 'input'; seq: number; up: boolean; down: boolean; left: boolean; right: boolean }

export type ServerMessage =
  | { type: 'welcome'; playerId: number; serverTime: number }
  | { type: 'pong'; clientTime: number; serverTime: number }
  | { type: 'state'; serverTime: number; players: PlayerSnapshot[] }
  | { type: 'error'; message: string }

export type PlayerSnapshot = {
  playerId: number
  x: number
  z: number
}

export function encodeClientMessage(message: ClientMessage): string {
  return JSON.stringify(message)
}

export function decodeServerMessage(line: string): ServerMessage {
  const value = JSON.parse(line) as unknown
  if (!isServerMessage(value)) {
    throw new Error(`Invalid server message: ${line}`)
  }
  return value
}

function isServerMessage(value: unknown): value is ServerMessage {
  if (!value || typeof value !== 'object') {
    return false
  }

  const message = value as Partial<ServerMessage>

  if (message.type === 'welcome') {
    return typeof message.playerId === 'number' && typeof message.serverTime === 'number'
  }

  if (message.type === 'pong') {
    return typeof message.clientTime === 'number' && typeof message.serverTime === 'number'
  }

  if (message.type === 'state') {
    return typeof message.serverTime === 'number' && Array.isArray(message.players) && message.players.every(isPlayerSnapshot)
  }

  if (message.type === 'error') {
    return typeof message.message === 'string'
  }

  return false
}

function isPlayerSnapshot(value: unknown): value is PlayerSnapshot {
  if (!value || typeof value !== 'object') {
    return false
  }

  const player = value as Partial<PlayerSnapshot>
  return typeof player.playerId === 'number' && typeof player.x === 'number' && typeof player.z === 'number'
}

File: server/src/game/protocol.rs
#[derive(Debug, PartialEq)]
pub enum ClientMessage {
    Join,
    Restart,
    Ping {
        ping_seq: u64,
    },
    Input {
        input_seq: u64,
        up: bool,
        down: bool,
        left: bool,
        right: bool,
    },
}

#[derive(Clone, Debug, PartialEq)]
pub enum ServerMessage {
    Welcome { player_id: u64 },
    Restarted,
    Pong { ping_seq: u64 },
    Error { message: String },
}

#[derive(Clone, Debug)]
pub struct PlayerSnapshot {
    pub player_id: u64,
    pub x: f32,
    pub y: f32,
    pub z: f32,
    pub vx: f32,
    pub vy: f32,
    pub vz: f32,
}

#[derive(Clone, Debug)]
pub struct BoxSnapshot {
    pub box_id: u64,
    pub x: f32,
    pub y: f32,
    pub z: f32,
    pub qx: f32,
    pub qy: f32,
    pub qz: f32,
    pub qw: f32,
    pub vx: f32,
    pub vy: f32,
    pub vz: f32,
    pub wx: f32,
    pub wy: f32,
    pub wz: f32,
}
